br/pkg/lightning/common/util.go

```diff
@@ -64,15 +68,42 @@ func (param *MySQLConnectParam) ToDSN() string {
 	return dsn
 }
 
-func (param *MySQLConnectParam) Connect() (*sql.DB, error) {
-	db, err := sql.Open("mysql", param.ToDSN())
+func (param *MySQLConnectParam) ToDSN() string {
+	return param.toDSN(param.Password)
+}
+
+func (param *MySQLConnectParam) tryConnect(password string) (*sql.DB, error) {
+	driverName := "mysql"
+	failpoint.Inject("MockMySQLDriver", func(val failpoint.Value) {
+		driverName = val.(string)
+	})
+	db, err := sql.Open(driverName, param.toDSN(password))
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
 
 	return db, errors.Trace(db.Ping())
 }
 
+func (param *MySQLConnectParam) Connect() (*sql.DB, error) {
+	db, err := param.tryConnect(param.Password)
+	if err == nil {
+		return db, nil
+	}
+	// If access is denied and password is encoded by base64, try the decoded string as well.
+	if mysqlErr, ok := errors.Cause(err).(*mysql.MySQLError); ok && mysqlErr.Number == tmysql.ErrAccessDenied {
+		// If password is encoded by base64, try the decoded string as well.
+		if password, decodeErr := base64.StdEncoding.DecodeString(param.Password); decodeErr == nil {
```

Comment:
How about try base decode first, if success, we try decoded password to connect first and if failed, then try the origin one.

Reply:
If password only contains digits, it always can be decoded as base64 string. So to keep consistent with previous behavior and avoid unsuccessfully database connect, here always tries plain password first.